### DISCUSSION TITLE
Need advice on handling multiple user roles

### DISCUSSION THREAD

**Discussion by VtorovN** [serverpod/serverpod#978] [Q&A] [answered]
Hello!

I'm creating an application in which User can have multiple roles and switch between them while not logging out of an account.

I've managed to embed UserRole enum to Authentication module, which now includes saving all available UserRoles to scopes. That works quite well, but with that I need to manage current UserRole on frontend and pass it to every backend method that has role separation.

So, now I'm trying to add chat module. Chats must be separated by UserRole too, e.g. User logged in with role "Foo" shouldn't see his own chats that belong to role "Bar". That seems like a big problem to me because I don't see a way to pass UserRole to get only chats for corresponding role.

The only thing I could come up with is copying the whole serverpod package to add some field like `currentRole` to `Session` to be able to get current UserRole directly from session as I usually get `authenticatedUserId`. With that, I would not need to manage UserRole on frontend anymore.

I'm asking for advice and ready for big refactor. Maybe there is some predefined mechanism, that I've missed, or just some better way.

**Comment by Frank1016** [comment]
Personally, I would do this: adding a Role field to Chats table, filtering chats based on current role, and updating the layout once role change detected.

**Reply by VtorovN** [reply]
The thing is - users with different roles can communicate with each other. Role Foo can chat with role Bar and vice versa. So adding Role to Chats table won't help - users with different roles can participate in one chat.

**Comment by Isakdl** [comment]
I'm not sure I fully understand the use-case. It seems you want to toggle what chats are shown to the user, but I don't follow why you need roles for this? Could you expand a little more on what you are trying to achieve? 

The scope system is a security feature meant for Authorization and normally you wouldn't let users swap between these as the scopes are there to limit access to resources in your system.

**Reply by VtorovN** [reply]
Basically the app is for communications between doctors and patients. So, user can be both doctor and patient at the same time with the same account, but the app functionality (including chats) must be different for the roles. Doctor may have dialogs with many patients, patients may have dialogs with many doctors. Therefore, if you are logged in as patient, you wouldn't want to see your own chats for doctor role.

It seems like I didn't fully understand what Scopes are for. Would you say that I need to implement roles somehow differently?

**Reply by Isakdl** [reply, marked as answer]
I see. I would not use the scopes to toggle what chats are visible, it is probably a good idea to use scopes here as well though! I presume doctors should be able to do things in the system a normal patient shouldn't be able to do. And you can limit this access in the API by using scopes. 

However for your specific problem with showing different chats depending on if you are in "doctor" or "patient" mode in the app. I think I would model this outside of the chats them selves. Having another entity that connects a doctor and a patient where you know if the current user is a doctor or a patient. 

Example:

```yaml
### use a better name!
class: DPConnection
table: dp_connection
fields:
  patientId: String,
  doctorId: String,
```

Then you can attach chats to this entity, as well as any other interesting data you want, journals, logs, whatever.

This way you would be able to fetch the user's connection by doctorId or by patientId depending on what mode you are in, in the app.

```dart
// as a patient
var connections = await DPConnection.find(
  session,
  where: (t) => t.patientId.equals(userId),
  limit 50,
);
```

```dart
// as a doctor
var connections = await DPConnection.find(
  session,
  where: (t) => t.doctorId.equals(userId),
  limit 50,
);
```

There are of course many ways to solve this, but I would probably solve it like above and I think using the scopes for this is the wrong approach. Scopes are there to **_restrict access_** and should only ever be allowed to be modified by an admin in the system.

**Reply by VtorovN** [reply]
Wow, thanks! Quite an interesting approach. I guess I will try it out.

Thanks for your time :)
